### ISSUE TITLE
[k8s.io] Restart [Disruptive] should restart all nodes and ensure all nodes and pods recover

### ISSUE THREAD

**Issue by davidopp** [kubernetes/kubernetes#32960]

**Comment by davidopp** [comment]
failing in
https://k8s-testgrid.appspot.com/google-1.3-1.4-upgrade#gce-cvm-1.3-cvm-1.4-upgrade-master

**Comment by pwittrock** [comment]
@dchen1107 Please help @jingxu97 triage this issue.  This looks like it could be a real issue that we need to address.  Since you have gone through the process of triaging this test before, please give @jingxu97 lots of direction so she can work on the issue efficiently.

**Comment by jingxu97** [comment]
I check this test, it only failed once for today since Sep 15 when the test started to run. For today's failure, it because in the first step to check all 3 nodes are ready, one node failed in ready state

```
Condition Ready of node gke-jenkins-e2e-default-pool-88628952-jniz is false instead of true. Reason: NodeStatusUnknown, message: Kubelet stopped posting node status.
Sep 19 06:03:41.668: INFO: Node gke-jenkins-e2e-default-pool-88628952-jniz didn't reach desired Ready condition status (true) within 17.98131326s
```

 But I checked the kubelet log for the node around the same time, the node still was working normally
```
I0919 13:03:21.275531    1233 kubelet.go:2566] SyncLoop (ADD, "api"): "kube-proxy-gke-jenkins-e2e-default-pool-88628952-jniz_kube-system(70f095a3-7e69-11e6-8c6b-42010af0001f)"
I0919 13:03:29.676288    1233 conversion.go:128] failed to handle multiple devices for container. Skipping Filesystem stats
I0919 13:03:29.676327    1233 conversion.go:128] failed to handle multiple devices for container. Skipping Filesystem stats
I0919 13:03:29.881770    1233 server.go:959] GET /healthz: (38.762µs) 200 [[curl/7.47.1] 127.0.0.1:49814]
I0919 13:03:38.884327    1233 server.go:959] GET /healthz: (27.475µs) 200 [[Go-http-client/1.1] 127.0.0.1:49828]
```

Will check the master log to see what is going on

**Comment by pwittrock** [comment]
You can see it failing consistently here:

- https://k8s-testgrid.appspot.com/google-1.3-1.4-upgrade#gce-cvm-1.3-cvm-1.4-upgrade-cluster
- https://k8s-testgrid.appspot.com/google-1.3-1.4-upgrade#gce-cvm-1.3-cvm-1.4-upgrade-cluster-new

**Comment by dchen1107** [comment]
They consistently failed with the following errors:

```
11:58:11 ERROR: gcloud crashed (AttributeError): 'module' object has no attribute 'GetError'
11:58:11 
11:58:11 If you would like to report this issue, please run the following command:
11:58:11   gcloud feedback
11:58:11 Traceback (most recent call last):
11:58:11   File "//google-cloud-sdk/lib/gcloud.py", line 65, in <module>
11:58:11     main()
11:58:11   File "//google-cloud-sdk/lib/gcloud.py", line 61, in main
11:58:11     sys.exit(googlecloudsdk.gcloud_main.main())
11:58:11   File "/google-cloud-sdk/lib/googlecloudsdk/gcloud_main.py", line 130, in main
11:58:11     gcloud_cli.Execute()
11:58:11   File "/google-cloud-sdk/lib/googlecloudsdk/calliope/cli.py", line 737, in Execute
11:58:11     self._HandleAllErrors(exc, command_path_string, flag_collection)
11:58:11   File "/google-cloud-sdk/lib/googlecloudsdk/calliope/cli.py", line 715, in Execute
11:58:11     resources = args.calliope_command.Run(cli=self, args=args)
11:58:11   File "/google-cloud-sdk/lib/googlecloudsdk/calliope/backend.py", line 1538, in Run
11:58:11     resources = command_instance.Run(args)
11:58:11   File "/google-cloud-sdk/lib/surface/compute/rolling_updates/start.py", line 131, in Run
11:58:11     raise exceptions.HttpException(updater_util.GetError(error))
11:58:11 AttributeError: 'module' object has no attribute 'GetError'
```
Here are the facts I collected based on the test dashboard:

* They only failed with two jobs: kubernetes-e2e-gce-1.3-1.4-upgrade-cluster & kubernetes-e2e-gce-1.3-1.4-upgrade-cluster-new. According to gcloud support doc, above issue normally is caused by partial upgrades of gcloud sdk, etc. and suggested to run command: gcloud component update to fix the issue.
* On another hand, I noticed that the same tests are not failed on any other jobs, like gke, 1.4 test suites, head test suites, etc. This makes me believe a recent change causing such regression for upgrade cases. And I found https://github.com/kubernetes/kubernetes/pull/32779 also noticed that the new change removes step of cluster/gce/upgrade.sh -P for preparing, which actually invoke gcloud component update.

cc/ @wojtek-t @davidopp @pwittrock

**Comment by jingxu97** [comment]
I traced the workflow, and it seem like the change #32779 is not used in this test.
It failed on this line https://github.com/kubernetes/kubernetes/blob/0877816ec8eae4739980d4d4b530801075e7c37e/test/e2e/framework/nodes_util.go#L128

@mikedanese could you please help take a look at it? This is related to MIG rolling update

**Comment by mikedanese** [comment]
The rpc to update the mig is failing with:
```json
{
 "error": {
  "errors": [
   {
    "domain": "global",
    "reason": "conditionNotMet",
    "message": "There already exists an update in a non-terminal state.",
    "locationType": "header",
    "location": "If-Match"
   }
  ],
  "code": 412,
  "message": "There already exists an update in a non-terminal state."
 }
}
```
and
```console
$  gcloud alpha compute rolling-updates --project=kube-gce-upg-1-3-1-4-upg-clu-n --zone=us-central1-f list --format json
```
shows
```json
[
  {
    "actionType": "RECREATE",
    "creationTimestamp": "2016-09-15T11:44:51.336Z",
    "error": {
      "errors": [
        {
          "code": "INVALID_VALUE"
        }
      ]
    },
    "id": "56d9c0eb-d77e-41d7-bf1c-df0738520047",
    "instanceGroupManager": "https://www.googleapis.com/compute/v1/projects/kube-gce-upg-1-3-1-4-upg-clu-n/zones/us-central1-f/instanceGroupManagers/jenkins-e2e-minion-group",
    "instanceTemplate": "https://www.googleapis.com/compute/v1/projects/kube-gce-upg-1-3-1-4-upg-clu-n/global/instanceTemplates/Instance template name: jenkins-e2e-minion-template-v1-4-0-beta-5",
    "kind": "replicapoolupdater#rollingUpdate",
    "oldInstanceTemplate": "https://www.googleapis.com/compute/v1/projects/kube-gce-upg-1-3-1-4-upg-clu-n/global/instanceTemplates/jenkins-e2e-minion-template",
    "policy": {
      "instanceStartupTimeoutSec": 300,
      "maxNumConcurrentInstances": 1
    },
    "progress": 0,
    "selfLink": "https://www.googleapis.com/replicapoolupdater/v1beta1/projects/kube-gce-upg-1-3-1-4-upg-clu-n/zones/us-central1-f/rollingUpdates/56d9c0eb-d77e-41d7-bf1c-df0738520047",
    "status": "PAUSED",
    "statusMessage": "0/3 instances updated; Errors: INVALID_VALUE",
    "user": "kubekins@kubernetes-jenkins.iam.gserviceaccount.com"
  }
]
```

**Comment by jingxu97** [comment]
So based on @mikedanese 's investigate, there are two issues.
1. gcloud alpha rolling updates crashes on any rpc error. Because of this, we could not get the real error message directly. File a bug against it  https://buganizer.corp.google.com/issues/31629319
2. On Sep 15th,  when the test first started, it tries to rolling update the nodes which failed because there is an existing update. But when trying to look this issue, the returned information is not very useful "code : INVALID_VALUE. Contact img team to investigate

Now we cancelled the failed update, and the status showed that ROLLED_UPDATE, will wait to see whether the test can pass after this.

**Comment by dchen1107** [comment]
This is a dup of #33113

**Comment by dchen1107** [comment]
cc/ @spxtr Can we fix the jenkin project first to unblock our release, then address #33113 in a proper way.

**Comment by spxtr** [comment]
Yes, cancel the rolling update and it should start passing. I'm looking into fixing the test.

**Comment by davidopp** [comment]
BTW I hit the same problem when I was doing manual upgrade testing a week ago. It's a very non-obvious failure mode (I did eventually figure it out).

**Comment by jingxu97** [comment]
The tests gce-cvm-1.3-cvm-1.4-upgrade-cluster-new started passing since yesterday after cancelling the pending update.
https://k8s-testgrid.appspot.com/google-1.3-1.4-upgrade#gce-cvm-1.3-cvm-1.4-upgrade-cluster-new&include-filter-by-regex=%5EOverall%24%7C%5C%5Bk8s%5C.io%5C%5D%5C%20Restart%5C%20%5C%5BDisruptive%5C%5D%5C%20should%5C%20restart%5C%20all%5C%20nodes%5C%20and%5C%20ensure%5C%20all%5C%20nodes%5C%20and%5C%20pods%5C%20recover

Will try to update  gce-cvm-1.3-cvm-1.4-upgrade-cluster too.

**Comment by dchen1107** [comment]
@jingxu97 I saw both *cluster-new and *cluster being passed since @psxtr cancelled the rolling update:

https://k8s-testgrid.appspot.com/google-1.3-1.4-upgrade#gce-cvm-1.3-cvm-1.4-upgrade-cluster&include-filter-by-regex=%5EOverall%24%7C%5C%5Bk8s%5C.io%5C%5D%5C%20Restart%5C%20%5C%5BDisruptive%5C%5D%5C%20should%5C%20restart%5C%20all%5C%20nodes%5C%20and%5C%20ensure%5C%20all%5C%20nodes%5C%20and%5C%20pods%5C%20recover

I am closing the issue. Gradefully handling restart node test cases is going to be addressed separately through #33113
